File: front/lib/front/memory_cookie.ex
defmodule Front.MemoryCookie do
  def values(nil) do
    defaults()
  end

  def values(cookie) do
    case Base.decode64(cookie) do
      {:ok, data} ->
        Map.merge(defaults(), Poison.decode!(data))

      _ ->
        defaults()
    end
  end

  defp defaults do
    %{
      "rootSidebar" => false,
      "rootRequester" => true,
      "projectType" => "",
      "projectListing" => "all_pipelines",
      "projectRequester" => "false",
      "logDark" => false,
      "logWrap" => true,
      "logLive" => true,
      "logSticky" => true,
      "logTimestamps" => true
    }
  end
end

File: front/assets/js/memory_cookie.js
import { Cookie } from "./cookie"

export class MemoryCookie {
  static set(key, value) {
    let values = this.getAll()
    values[key] = value

    values = JSON.stringify(values)
    values = window.btoa(values)

    Cookie.setPermanent('memory', values, false)
  }

  static get(key) {
    let values = this.getAll()

    return values[key]
  }

  static getAll() {
    let values = Cookie.get('memory')

    if(values == null) {
      values = {}
    } else {
      values = window.atob(values)
      values = JSON.parse(values)
    }

    let defaults = {
      rootSidebar: false,
      rootRequester: true,
      projectType: "",
      projectListing: "all_pipelines",
      projectRequester: "false",
      logDark: false,
      logWrap: true,
      logLive: true,
      logSticky: true,
      logTimestamps: true
    };

    return Object.assign(defaults, values || {});
  }
}
